### ISSUE TITLE
Weird audio issues with high audio per packet settings

### ISSUE THREAD

**Issue by bkacjios** [mumble-voip/mumble#6415]
### Description

Hello, I am the author of lua-mumble, a module for allowing someone to create a bot via the Lua scripting language.

I've been updating it to support the new protobuf UDP packets that were introduced in 1.5 and have been running into this weird issue. I have a test bot that is configured with a 60 ms audio frame per packet with a 96000 bitrate. It is also configured to output stereo audio. The server has a max bitrate of 558000.

I noticed that sometimes I was having weird audio issues with the bot I couldn't explain. Sometimes I would start the bot and the audio would be stuttering from the start, other times I would start it and it would sound perfect. The only way I was able to reproduce it consistently is by setting the audio packet size to 60 on the bot and muting/unmuting my client.

https://github.com/mumble-voip/mumble/assets/3247233/d4b7a660-76a1-4da9-999c-9128223647fa

I even managed to get this to happen when having my bot not play any music at all. Instead I had it loop back my audio so I could hear myself through the bot. The stuttering issues were also present. It almost seems like my clients decoder is trying to decode it in some other mode or something, but honestly I'm not too sure what is going on here. 

Would anyone perhaps know what is going on here? I'm still unsure if this is an issue with the way I am encoding audio or an issue with how mumble is decoding audio.

### Steps to reproduce

Have a client start transmitting audio at 60ms audio frame packet.
Mute and unmute the audio. They will now be stuttering.

### Mumble version

1.5.628

### Mumble component

Client

### OS

Windows

### Reproducible?

Yes

**Comment by davidebeatrici** [comment]
I have encountered a similar issue when working on a project based on libmumble.

Turns out the Mumble client struggles with packets that contain more than 10ms (480 samples at 48000 Hz) of audio.

Try to send 480 samples (per channel): the issue should not be present.

**Comment by bkacjios** [comment]
Good to know. I was kinda going crazy thinking I was doing something wrong somewhere. Yeah, setting it to 10 ms seems to work fine. 20 is okay too mostly, but 40 and 60 are basically unusable. I usually like to use at least 20 ms since I find it seems to keep the audio clarity stable. I notice little hiccups or stutters with 10 that are nowhere near as bad as this though.

**Comment by davidebeatrici** [comment]
Thank you for confirming.

This is definitely something that has to be fixed in the client. Ideally it should handle all frame sizes that are supported by libopus.

https://github.com/mumble-voip/mumble/blob/dfc8dacae3f37ecd522d73fe7db9a0a91060e0ae/src/mumble/Audio.h#L22
https://github.com/mumble-voip/mumble/blob/dfc8dacae3f37ecd522d73fe7db9a0a91060e0ae/src/mumble/AudioInput.h#L219-L224

The `iSampleRate` and `iFrameSize` variables exist because the code itself doesn't consider the sample rate and frame size to be hardcoded.

Unfortunately an option to change them was never added and as a result problems like this one went unnoticed until third-party clients (e.g. bots) started popping up.

**Comment by bkacjios** [comment]
So wait, I'm a little confused. I thought this setting in the client adjusts it? Are you saying the client just thinks all incoming audio is 10 ms?

![image](https://github.com/mumble-voip/mumble/assets/3247233/2440b4a2-c30c-48b2-9357-3f2dfa6e3bc4)

I made my bot mimic this functionality. I was defaulting it to 20 ms like mumble does on a fresh install, but I can adjust it via a method call.

https://github.com/bkacjios/lua-mumble/blob/54c2272e2f0ba06bacf87c2fdc21441d14505140/mumble/mumble.h#L81
https://github.com/bkacjios/lua-mumble/blob/54c2272e2f0ba06bacf87c2fdc21441d14505140/mumble/audio.c#L196

**Comment by davidebeatrici** [comment]
The setting itself adjusts the number of 10ms chunks to send per packet, not the actual number of audio frames:

https://github.com/mumble-voip/mumble/blob/dfc8dacae3f37ecd522d73fe7db9a0a91060e0ae/src/mumble/AudioInput.h#L235-L239
https://github.com/mumble-voip/mumble/blob/dfc8dacae3f37ecd522d73fe7db9a0a91060e0ae/src/mumble/AudioInput.cpp#L694-L728
https://github.com/mumble-voip/mumble/blob/dfc8dacae3f37ecd522d73fe7db9a0a91060e0ae/src/mumble/AudioInput.cpp#L1105-L1140

Another issue in the code above is that it doesn't clamp the number of chunks to match `iAudioFrame`'s value.

Putting that aside, let's see what is going on in the audio output section:

https://github.com/mumble-voip/mumble/blob/dfc8dacae3f37ecd522d73fe7db9a0a91060e0ae/src/mumble/AudioOutputSpeech.cpp#L58-L153

As you can see, the code assumes that incoming frames are always 10ms. This is wrong because (at least in our case) the Opus encoder always produces packets that contain a single encoded frame.

Basically, there is no concept of "chunks" in encoded packets, regardless of the client's audio input settings.

Finally, just to add some more confusion to the mix and possibly clarifying it: xiph/opus#315

**Comment by bkacjios** [comment]
Ahhh.. I understand now. My bot has a timer that determines how often it should send the audio data. So if I set the bots audio packet size to 60, it's encoding 2880 frames into one audio packet every 60 ms. I'm guessing I should send 6 individual packets with 480 encoded frames in one go?

**Comment by davidebeatrici** [comment]
Yup!

**Comment by bkacjios** [comment]
Am I missing something?

When my bot receives audio from me speaking, it isn't getting 10 ms chunks if I adjust my client to have a 60 ms delay. Doesn't this go against what you said? Shouldn't I still be receiving 6 individual 240 bytes every 0.06 seconds?

```[MUMBLE - TRACE] RECEIVED MumbleUDP.Audio
OnUserStartSpeaking     mumble.user [137]["Bkacjios"]
received 960 bytes in 0.059990 seconds
[MUMBLE - TRACE] RECEIVED MumbleUDP.Audio
received 960 bytes in 0.059892 seconds
[MUMBLE - TRACE] RECEIVED MumbleUDP.Audio
received 960 bytes in 0.060152 seconds
OnUserStopSpeaking      mumble.user [137]["Bkacjios"]
```

Compared to the results when I have it set to 10 ms.

```
[MUMBLE - TRACE] RECEIVED MumbleUDP.Audio
OnUserStartSpeaking     mumble.user [137]["Bkacjios"]
received 240 bytes in 0.009980 seconds
[MUMBLE - TRACE] RECEIVED MumbleUDP.Audio
received 240 bytes in 0.009977 seconds
[MUMBLE - TRACE] RECEIVED MumbleUDP.Audio
received 240 bytes in 0.009891 seconds
OnUserStopSpeaking      mumble.user [137]["Bkacjios"]
```

**Comment by davidebeatrici** [comment]
Are you referring to the size of the encoded packet or the raw audio data?

**Comment by bkacjios** [comment]
Yeah, this was the encoded packet I receive from a speaking client.

I'm starting to think we had a misunderstanding here.

> Turns out the Mumble client struggles with packets that contain more than 10ms (480 samples at 48000 Hz) of audio.
>
> Try to send 480 samples (per channel): the issue should not be present.

My bot resamples all playing audio to 48000hz, since that's what mumble expects. I always encode 480 samples per 10ms. The issue I am having is with sending more than 10 ms per packet. (Replicating the audio per packet setting in an official client)

If I set it to 20, I will encode (480 * 2) bytes of PCM data and send that over. Is that not correct?

**Comment by bkacjios** [comment]
After much investigation, I think this is more an issue with my bot than it is with mumble itself. [The timer system I am using to send out data is only accurate at a millisecond level and ends up causing issues.](https://docs.libuv.org/en/v1.x/timer.html)

These debug logs are from my bot client's audio timer. Transmission time is how long it took me to process any playing audio files or streams and to send it out over the network.

```
[MUMBLE - DEBUG] audio transmission took: 0.342 ms
[MUMBLE - DEBUG] next timer length: 9 ms
[MUMBLE - DEBUG] audio transmission took: 0.466 ms
[MUMBLE - DEBUG] next timer length: 9 ms
[MUMBLE - DEBUG] audio transmission took: 0.345 ms
[MUMBLE - DEBUG] next timer length: 9 ms
[MUMBLE - DEBUG] audio transmission took: 0.383 ms
[MUMBLE - DEBUG] next timer length: 9 ms
[MUMBLE - DEBUG] audio transmission took: 0.428 ms
[MUMBLE - DEBUG] next timer length: 9 ms
```

So when sending out a 10ms audio packet, it's having to run the audio timer every 9 milliseconds to keep up. This is obviously too fast for mumbles liking, so the mumble client audio skips when it tries to keep up.

```
[MUMBLE - DEBUG] audio transmission took: 0.604 ms
[MUMBLE - DEBUG] next timer length: 19 ms
[MUMBLE - DEBUG] audio transmission took: 0.603 ms
[MUMBLE - DEBUG] next timer length: 19 ms
[MUMBLE - DEBUG] audio transmission took: 0.583 ms
[MUMBLE - DEBUG] next timer length: 19 ms
[MUMBLE - DEBUG] audio transmission took: 0.547 ms
[MUMBLE - DEBUG] next timer length: 19 ms
[MUMBLE - DEBUG] audio transmission took: 0.363 ms
[MUMBLE - DEBUG] next timer length: 19 ms
```

At 20ms, things are much better, and everything works fine. The 19 millisecond timer ends up working out for me here and ends up being a best case scenario for the CPU I am running it on.

At 40ms, things start to go downhill again.

```
[MUMBLE - DEBUG] audio transmission took: 0.710 ms
[MUMBLE - DEBUG] next timer length: 39 ms
[MUMBLE - DEBUG] audio transmission took: 0.558 ms
[MUMBLE - DEBUG] next timer length: 39 ms
[MUMBLE - DEBUG] audio transmission took: 0.706 ms
[MUMBLE - DEBUG] next timer length: 39 ms
[MUMBLE - DEBUG] audio transmission took: 1.039 ms
[MUMBLE - DEBUG] next timer length: 38 ms
[MUMBLE - DEBUG] audio transmission took: 1.196 ms
[MUMBLE - DEBUG] next timer length: 38 ms
```

It ends up jumping around between running the audio loop every 38-39 milliseconds this time, causing inconsistencies.

At 60ms, things get better again. Sort of. The audio will still sound like it did in the video in the OP for a moment, but it will usually settle down and start sounding fine again.

```
[MUMBLE - DEBUG] audio transmission took: 1.867 ms
[MUMBLE - DEBUG] next timer length: 58 ms
[MUMBLE - DEBUG] audio transmission took: 1.829 ms
[MUMBLE - DEBUG] next timer length: 58 ms
[MUMBLE - DEBUG] audio transmission took: 1.856 ms
[MUMBLE - DEBUG] next timer length: 58 ms
[MUMBLE - DEBUG] audio transmission took: 1.930 ms
[MUMBLE - DEBUG] next timer length: 58 ms
[MUMBLE - DEBUG] audio transmission took: 1.841 ms
[MUMBLE - DEBUG] next timer length: 58 ms
[MUMBLE - DEBUG] audio transmission took: 1.829 ms
[MUMBLE - DEBUG] next timer length: 58 ms
```

I think it all just comes down to mumble being very picky about how fast audio data is allowed to be streamed in. I think I need to rewrite my bot to buffer audio data ahead of time, in a separate thread. Then I can just have a timer just reading and sending out that buffered data out. The CPU processing time is just too inconsistent.

**Comment by Krzmbrzl** [comment]
Thanks for reporting back with your findings 👍 

> I think it all just comes down to mumble being very picky about how fast audio data is allowed to be streamed in.

Unfortunately, the audio handling code is a bit in a bad-ish shape and has multiple hard-wire assumptions and required coded into it. There is hope that things will improve with the switch to libcrossaudio and/or libmumble but those are more of long-term goals and nothing that will happen in the near future.
